languages/js/src/filter.ts

```diff
@@ -0,0 +1,140 @@
+import type { Host } from './Host';
+import type { PolarComparisonOperator, PolarValue, obj } from './types';
+import { OsoError } from './errors';
+
+export interface SerializedRelation {
+  Relation: {
+    kind: string;
+    other_class_tag: string;
+    my_field: string;
+    other_field: string;
+  };
+}
+
+type RelationKind = 'one' | 'many';
+
+/** Represents relationships between two resources, eg. one-one or one-many. */
+export class Relation {
+  kind: RelationKind;
+  otherType: string;
+  myField: string;
+  otherField: string;
+
+  constructor(
+    kind: RelationKind,
+    otherType: string,
+    myField: string,
+    otherField: string
+  ) {
+    this.kind = kind;
+    this.otherType = otherType;
+    this.myField = myField;
+    this.otherField = otherField;
+  }
+
+  serialize(): SerializedRelation {
+    return {
+      Relation: {
+        kind: this.kind,
+        other_class_tag: this.otherType,
+        my_field: this.myField,
+        other_field: this.otherField,
+      },
+    };
+  }
+}
+
+export type SerializedFields = {
+  [field: string]: SerializedRelation | { Base: { class_tag: string } };
+};
+
+interface FilterRelation {
+  fromTypeName: string;
+  fromFieldName: string;
+  toTypeName: string;
+}
+
+export interface Projection {
+  typeName: string;
+  fieldName: string;
+}
+
+export interface Immediate {
+  value: unknown;
+}
+
+export interface Adapter<Query, Resource> {
+  buildQuery: (f: Filter) => Query;
+  executeQuery: (q: Query) => Promise<Resource[]>;
+}
+
+export type Datum = Projection | Immediate;
+
+export interface FilterCondition {
+  lhs: Datum;
+  cmp: PolarComparisonOperator;
+  rhs: Datum;
+}
+
+export type FilterConditionSide = 'lhs' | 'rhs';
+
+export interface Filter {
+  model: string;
+  relations: FilterRelation[];
+  conditions: FilterCondition[][];
+  types: { [tag: string]: SerializedFields };
+}
+
+export interface FilterJson {
+  conditions: unknown[][][];
+  relations: string[][];
+  root: string;
+}
```

Comment:
Comments (a la https://github.com/osohq/oso/blob/f75566c1a885966646741a43029524b0a283e023/languages/js/src/filter.ts#L16) for at least some of these types would be helpful. E.g., `FilterConditionSide` is pretty obvious but `Projection` & `Immediate` aren't

Reply:
added in d3d40b5